kernel.c

```diff
@@ -0,0 +1,278 @@
+    }
+
+    // Set up handlers for each trap type by populating the vector table
+    // Each entry contains the address of the function to handle that specific trap
+    vector_table[TRAP_KERNEL] = (int)kernel_handler;        // System calls from user processes
+    vector_table[TRAP_CLOCK] = (int)clock_handler;          // Timer interrupts for scheduling
+    vector_table[TRAP_ILLEGAL] = (int)illegal_handler;      // Illegal instruction exceptions
+    vector_table[TRAP_MEMORY] = (int)memory_handler;        // Memory access violations and stack growth
+    vector_table[TRAP_MATH] = (int)math_handler;            // Math errors like division by zero
+    vector_table[TRAP_TTY_RECEIVE] = (int)receive_handler;   // Terminal input available
+    vector_table[TRAP_TTY_TRANSMIT] = (int)transmit_handler; // Terminal output complete
+
+    // Write the address of vector table to REG_VECTOR_BASE register
+    // This tells the hardware where to find the interrupt handlers
+    WriteRegister(REG_VECTOR_BASE, (unsigned int)vector_table);
+
+    TracePrintf(0, "Interrupt vector table initialized at 0x%p\n", vector_table);
+}
+
+pcb_t *create_idle_process(UserContext *uctxt) {
+    TracePrintf(1, "ENTER create_idle_process.\n");
+
+    // Create a new PCB for the idle process
+    pcb_t *idle_pcb = create_pcb();
+    if (idle_pcb == NULL) {
+        TracePrintf(0, "ERROR: Failed to allocate idle PCB\n");
+        return NULL;
+    }
+
+    // Get the current Region 1 page table pointer - this is a physical address at this point
+    unsigned int r1_page_table_addr = ReadRegister(REG_PTBR1);
+
+    // After VM is enabled, you'll need to access this through virtual memory
+    // Store the physical address for now - your kernel will need to map this later
+    idle_pcb->region1_pt = (struct pte *)r1_page_table_addr;
+
+    // For kernel stack, save the page frame numbers
+    // Since you're mapping 1:1 with page = frame, save these page numbers
+    int base_kernelStack_page = KERNEL_STACK_BASE / PAGESIZE;
+    int kernelStack_limit_page = KERNEL_STACK_LIMIT / PAGESIZE;
+
+    // Allocate space to store kernel stack frame numbers
+    int num_stack_frames = kernelStack_limit_page - base_kernelStack_page;
+    idle_pcb->kernel_stack_pages = malloc(num_stack_frames * sizeof(int));
+    if (idle_pcb->kernel_stack_pages == NULL) {
+        TracePrintf(0, "ERROR: Failed to allocate space for kernel stack frames\n");
+        free(idle_pcb);
+        return NULL;
+    }
+
+    // Save the frame numbers for the kernel stack
+    for (int i = 0; i < num_stack_frames; i++) {
+        idle_pcb->kernel_stack_pages[i] = base_kernelStack_page + i;
+    }
+
+    // Copy the UserContext provided to KernelStart
+    memcpy(&idle_pcb->user_context, uctxt, sizeof(UserContext));
+
+    // Point PC to DoIdle function and SP to top of user stack
+    idle_pcb->user_context.pc = (void *)DoIdle;
+    idle_pcb->user_context.sp = (void *)VMEM_1_LIMIT;  // Top of Region 1

```

Comment:
this is a bit tricky -- you actually need to offset by 4 bytes to allow for the initial stack write (of an integer I believe. I'm not sure where it's documented). so you can do `idle_pcb->user_context.sp = (void *) VMEM_1_LIMIT - 4